### PR TITLE
Add Compressor

### PR DESCRIPTION
**Pull request by KAIYOHUGO**
Impl basic compressor #32

### CODE REVIEW

File: examples/leveldb-tool/src/main.rs [hunk 1]
```diff
@@ -59,7 +59,7 @@ fn main() {
     let mut opt = Options::default();
     opt.reuse_logs = false;
     opt.reuse_manifest = false;
-    opt.compression_type = rusty_leveldb::CompressionType::CompressionSnappy;
+    opt.compressor = rusty_leveldb::compressor::SnappyCompressor::ID;
```
Comment (dermesser) [inline examples/leveldb-tool/src/main.rs:62]: cam you import `compressor`? (yes I know, previously this also wasn't done)

File: src/error.rs [hunk 2]
```diff
@@ -20,6 +20,7 @@ pub enum StatusCode {
     IOError,
     InvalidArgument,
     InvalidData,
+    InvalidOption,
     LockError,
     NotFound,
     NotSupported,
```
Comment (dermesser) [inline src/error.rs:26]: this might be the appropriate error code for unsupported compression types.

File: src/options.rs [hunk 3]
```diff
@@ -51,14 +38,21 @@ pub struct Options {
     pub block_restart_interval: usize,
     /// Note: you have to open a database with the same compression type as it was written to, in
     /// order to not lose data! (this is a bug and will be fixed)
-    pub compression_type: CompressionType,
+    pub compressor: u8,
+
+    pub compressor_list: Rc<[Option<Box<dyn Compressor>>; 256]>,
```
Comment (dermesser) [inline src/options.rs:43]: this is a good idea, but could you encapsulate this? for example with an API like

```rust
let mut compressor_list = CompressorList::new();
compressor_list.set(/*type =*/0, NoneCompressor::new());
compressor_list.set(1, SnappyCompressor::new());
let c = compressor_list.get(1);
```

That would look a bit nicer to users of the `Options` type supplying their own compressors.

**Review by dermesser — CHANGES_REQUESTED** [review]
Thank you very much for this, it looks very good already! Also thanks for being patient with me.

File: src/options.rs (hunk 3 above)
Comment (KAIYOHUGO) [reply src/options.rs:43]: You are right! It look nicer and cleaner in user api and internal code.

**Comment by KAIYOHUGO** [comment]
Thank for your review! I fix the problem you mention above, and add a example about customize compressor.

File: src/compressor.rs [hunk 4]
```diff
@@ -1,22 +1,46 @@
+/// Custom compression method
+///
+/// ```
+/// # use rusty_leveldb::{Compressor, CompressorId};
+/// 
+/// #[derive(Debug, Clone, Copy, Default)]
+/// pub struct CustomCompressor;
+///
+/// impl CompressorId for CustomCompressor {
+///     const ID: u8 = 0;
```
Comment (dermesser) [inline src/compressor.rs:10]: a small detail: given that `0` is for the "none" compressor, maybe use an obviously bogus number like "42"

File: src/options.rs [hunk 5]
```diff
@@ -70,12 +67,64 @@ impl Default for Options {
...
+        self.0[id as usize] = Some(Box::new(compressor));
+    }
+
+    pub fn is_set(&self, id: u8) -> bool {
+        self.0[id as usize].is_some()
+    }
+
+    pub fn get(&self, id: u8) -> Result<&Box<dyn Compressor + 'static>> {
+        self.0[id as usize].as_ref().ok_or_else(|| Status {
+            code: StatusCode::InvalidData,
```
Comment (dermesser) [inline src/options.rs:113]: another detail: maybe `NotSupported` is more appropriate?

File: src/table_block.rs [hunk 6]
```diff
@@ -70,11 +70,11 @@ pub fn read_table_block(
...
     let compressor_list = opt.compressor_list.clone();
-    if let Some(compressor) = compressor_list[compress[0] as usize].as_ref() {
-        Ok(Block::new(opt, compressor.decode(buf)?))
-    } else {
-        err(StatusCode::InvalidData, "invalid compression type")
-    }
+
+    Ok(Block::new(
+        opt,
+        compressor_list.get(compress[0])?.decode(buf)?,
```
Comment (dermesser) [inline src/table_block.rs:76]: beautiful! :-)

File: examples/mcpe/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+    opt
+}
+
+fn main() {
+    // let path = "path here";
+    // let compression_level = 10;
+    // let opt = mcpe_options(compression_level);
+    // DB::open(path, opt)
+
+    // Do something
```
Comment (dermesser) [inline examples/mcpe/src/main.rs:68]: maybe you can just write one or two entries (and read them back), to serve as a basic sanity check.

**Review by dermesser — CHANGES_REQUESTED** [review]
Almost done - thank you very much, this is great work!

File: src/compressor.rs (hunk 4 above)
Comment (KAIYOHUGO) [reply src/compressor.rs:10]: Oh I forget to change that. I ctrl-c ctrl-v that piece of code from NoneCompressor. 😅

**Comment by KAIYOHUGO** [comment]
Done :tada:

**Review by dermesser — APPROVED** [review]

**Pull request merged**